### PR TITLE
Field updates, strategy abstraction, auton refactoring

### PR DESCRIPTION
**Pull request by billwpierce**
1. Changes the organization of the field with javadocs, updates to the piece, and better constructors.
2. Adds a strategy layer on top of commandgroup in case we need to initialize without constructing
3. Refactors auton calls to work without immediate construction.

### CODE REVIEW

File: src/org/usfirst/frc4904/autonly/CenterSwitchDistance.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package org.usfirst.frc4904.autonly;
+
+
+import org.usfirst.frc4904.robot.RobotMap;
+import org.usfirst.frc4904.standard.commands.RunIfElse;
+import org.usfirst.frc4904.standard.commands.chassis.ChassisMoveDistance;
+import org.usfirst.frc4904.standard.commands.chassis.ChassisTurn;
+
+public class CenterSwitchDistance extends Strategy {
+	public static final double DISTANCE_APPROACH_MID = -1; // TODO: real values
```
Comment (oldgalileo) [inline src/org/usfirst/frc4904/autonly/CenterSwitchDistance.java:10]: There's no better time "TODO" like the present. Get these values and test them.

File: src/org/usfirst/frc4904/autonly/CenterSwitchDistance.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+	public static final double DISTANCE_MID_SWITCH = -1;
+	public static final double DEGREES_TURN_LEFT = -1;
+	public static final double DEGREES_TURN_RIGHT = -1;
+
+	public CenterSwitchDistance() {
+		addSequential(
+			new ChassisMoveDistance(RobotMap.Component.chassis, RobotMap.Metrics.Wheel.TICKS_PER_INCH * DISTANCE_APPROACH_MID,
```
Comment (oldgalileo) [inline src/org/usfirst/frc4904/autonly/CenterSwitchDistance.java:17]: I'm like 90% sure that our formatter doesn't opt for a newline after the parenthesis on a method call. Please fix.

File: src/org/usfirst/frc4904/autonly/Field.java [hunk 3]
```diff
@@ -1,25 +1,83 @@
...
 
 public class Field {
 	public Piece ourSwitch;
 	public Piece scale;
 	public Piece theirSwitch;
-	public Alliance ourTeam;
-
+	public Alliance team;
+	
+	/** The full constructor of the field.
```
Comment (oldgalileo) [inline src/org/usfirst/frc4904/autonly/Field.java]: The JavaDocs are really good, however it's the norm to not start the documentation on the first line. Most people opt for something like:
```java
/**
 * The full constructor of the field.
 *
 * @param
 * @param
 */
```

File: src/org/usfirst/frc4904/autonly/Field.java [hunk 4]
```diff
@@ -29,5 +87,36 @@ public Piece(Alliance team, char reference) {
...
+		public String toString() {
+			return "Piece{L:'" + left.name() + "' R:'"+ right.name() +"'}";
+		}
+		
+		/** Tells us if the left side of the pieces is ours.
+		 * 
+		 * @return true if the left side is ours, false if it is not.
+		 */
+		public boolean isLeftOurs(){
+			return (DriverStation.getInstance().getAlliance() == this.left) ? true : false;
```
Comment (oldgalileo) [inline src/org/usfirst/frc4904/autonly/Field.java]: `isLeftOurs()` and `isRightOurs()` should be changed to ensure that it's not comparing two Alliance.Invalid values. We don't want to run any side-based auton if the field didn't send us the proper field configuration value.

File: src/org/usfirst/frc4904/autonly/LeftSideDistance.java [hunk 5]
```diff
@@ -1,9 +1,13 @@
 package org.usfirst.frc4904.autonly;
 
 import org.usfirst.frc4904.robot.RobotMap;
+import org.usfirst.frc4904.standard.commands.Noop;
+import org.usfirst.frc4904.standard.commands.RunIf;
 
 public class LeftSideDistance extends SideSwitchDistance {
 	public LeftSideDistance () {
-		super(RobotMap.gamefield.ourSwitch.left);
+		addSequential(new RunIf(new Noop(), () -> {
```
Comment (oldgalileo) [inline src/org/usfirst/frc4904/autonly/LeftSideDistance.java]: Shouldn't be running a Noop, this should be running code that works.

File: src/org/usfirst/frc4904/autonly/LeftSideTime.java [hunk 6]
```diff
@@ -1,9 +1,13 @@
...
 
 import org.usfirst.frc4904.robot.RobotMap;
+import org.usfirst.frc4904.standard.commands.Noop;
+import org.usfirst.frc4904.standard.commands.RunIf;
 
 public class LeftSideTime extends SideSwitchTime {
 	public LeftSideTime () {
-		super(RobotMap.gamefield.ourSwitch.left);
-	}
+    	addSequential(new RunIf(new Noop(), () -> {
```
Comment (oldgalileo) [inline src/org/usfirst/frc4904/autonly/LeftSideTime.java]: Again, no Noop.

File: src/org/usfirst/frc4904/autonly/SideSwitchDistance.java [hunk 7]
```diff
@@ -1,20 +1,15 @@
...
-import edu.wpi.first.wpilibj.DriverStation.Alliance;
-import edu.wpi.first.wpilibj.command.CommandGroup;
 
-/**
- *
- */
-public class SideSwitchDistance extends CommandGroup {
-	public static final double DISTANCE_APPROACH_SWITCH = 9; // TODO: Test this ditance.
+public class SideSwitchDistance extends Strategy {
+	public static final double DISTANCE_APPROACH_SWITCH = 9; // TODO: Test this distance.
```
Comment (oldgalileo) [inline src/org/usfirst/frc4904/autonly/SideSwitchDistance.java:7]: No time like the present to test this distance.

**Review by oldgalileo — CHANGES_REQUESTED** [review]

File: src/org/usfirst/frc4904/autonly/LeftSideTime.java (hunk 6 above)
Comment (oldgalileo) [reply src/org/usfirst/frc4904/autonly/LeftSideTime.java]: There are a few other places this happens but I was too lazy to go around and add comments on each one. Just imagine I did and change them 👌

File: src/org/usfirst/frc4904/autonly/Field.java [hunk 8]
```diff
@@ -1,25 +1,83 @@
...
+	}
+		
+	@Override
+	public String toString() {
+		return "Field{Team:'"+team.name()+"' OurSwitch:'" + ourSwitch.toString() + "' Scale:'" + scale.toString() + "' TheirSwitch:'"+theirSwitch.toString()+"'}";
 	}
 
+	/** The piece object, which is either a switch or the scale. Contains the alliances of both the left and right side of the piece from the position of our alliance.
+	 * 
+	 * @author bilpier
```
Comment (ajnadel) [inline src/org/usfirst/frc4904/autonly/Field.java]: ...

File: src/org/usfirst/frc4904/autonly/Field.java (hunk 4 above)
Comment (ajnadel) [inline src/org/usfirst/frc4904/autonly/Field.java]: @HowardStark hey check this out

File: src/org/usfirst/frc4904/autonly/LeftSideDistance.java (hunk 5 above)
Comment (ajnadel) [inline src/org/usfirst/frc4904/autonly/LeftSideDistance.java]: Add an explicative comment if there is a TODO

File: src/org/usfirst/frc4904/autonly/LeftSideTime.java (hunk 6 above)
Comment (ajnadel) [inline src/org/usfirst/frc4904/autonly/LeftSideTime.java]: See previous.

File: src/org/usfirst/frc4904/autonly/RightSideDistance.java [hunk 9]
```diff
@@ -1,9 +1,13 @@
 package org.usfirst.frc4904.autonly;
 
 import org.usfirst.frc4904.robot.RobotMap;
+import org.usfirst.frc4904.standard.commands.Noop;
+import org.usfirst.frc4904.standard.commands.RunIf;
 
 public class RightSideDistance extends SideSwitchDistance {
 	public RightSideDistance () {
-		super(RobotMap.gamefield.ourSwitch.right);
+		addSequential(new RunIf(new Noop(), () -> {
```
Comment (ajnadel) [inline src/org/usfirst/frc4904/autonly/RightSideDistance.java]: .

File: src/org/usfirst/frc4904/autonly/RightSideTime.java [hunk 10]
```diff
@@ -1,9 +1,13 @@
...
 
 import org.usfirst.frc4904.robot.RobotMap;
+import org.usfirst.frc4904.standard.commands.Noop;
+import org.usfirst.frc4904.standard.commands.RunIf;
 
 public class RightSideTime extends SideSwitchTime {
 	public RightSideTime () {
-		super(RobotMap.gamefield.ourSwitch.right);
-	}
+    	addSequential(new RunIf(new Noop(), () -> {
```
Comment (ajnadel) [inline src/org/usfirst/frc4904/autonly/RightSideTime.java]: .

File: src/org/usfirst/frc4904/autonly/CenterSwitchDistance.java (hunk 2 above)
Comment (oldgalileo) [reply src/org/usfirst/frc4904/autonly/CenterSwitchDistance.java:17]: Apparently I was wrong which is worse.

File: src/org/usfirst/frc4904/autonly/Field.java [hunk 11]
```diff
@@ -29,5 +104,43 @@ public Piece(Alliance team, char reference) {
...
+			Alliance currentAlliance = DriverStation.getInstance().getAlliance();
+			return currentAlliance != Alliance.Invalid && currentAlliance == this.left;
+		}
+
+		/**
+		 * Tells us if the right side of the pieces is ours.
+		 * 
+		 * @return true if the right side is ours, false if it is not.
+		 */
+		public boolean isRightOurs() {
```
Comment (oldgalileo) [inline src/org/usfirst/frc4904/autonly/Field.java:141]: Looks good.

**Review by oldgalileo — APPROVED** [review]

File: src/org/usfirst/frc4904/autonly/SideSwitchDistance.java (hunk 7 above)
Comment (ajnadel) [inline src/org/usfirst/frc4904/autonly/SideSwitchDistance.java:7]: Include units in name?

File: src/org/usfirst/frc4904/autonly/SideSwitchTime.java [hunk 12]
```diff
@@ -1,21 +1,15 @@
...
- */
-public class SideSwitchTime extends CommandGroup {
-	public static final double TIME_APPROACH_SWITCH = -1; // TODO: Replace with real time
-
-    public SideSwitchTime(Alliance switchSideAlliance) { //TODO: Add rest of code.
-        addSequential(new ChassisConstant(RobotMap.Component.chassis, 0, AutonConfig.DEAD_RECKON_DRIVE_SPEED, 0,
-			SideSwitchTime.TIME_APPROACH_SWITCH));
-        addSequential(new RunIf(new Noop(), () -> switchSideAlliance == RobotMap.gamefield.ourTeam)); //TODO: Replace with outtake.
+public class SideSwitchTime extends Strategy {
+	public static final double TIME_APPROACH_SWITCH = 3; // TODO: Replace with real time
```
Comment (ajnadel) [inline src/org/usfirst/frc4904/autonly/SideSwitchTime.java:7]: Units

File: src/org/usfirst/frc4904/autonly/Strategy.java [hunk 13]
```diff
@@ -0,0 +1,7 @@
+package org.usfirst.frc4904.autonly;
+
+import edu.wpi.first.wpilibj.command.CommandGroup;
+
+public abstract class Strategy extends CommandGroup {
```
Comment (ajnadel) [inline src/org/usfirst/frc4904/autonly/Strategy.java:5]: I still maintain that this is useless.

**Pull request merged**